### PR TITLE
OnCastComplete rework

### PR DESCRIPTION
**Pull request by Tharre**

### CODE REVIEW

File: sim/druid/balance/TestBalance.results [hunk 1]
```diff
@@ -46,911 +46,911 @@ character_stats_results: {
...
  key: "TestBalance-AllItems-WrathfulGladiator'sIdolofResolve-51429"
  value: {
-  dps: 7355.53047
-  tps: 7194.44067
+  dps: 7320.79478
+  tps: 7159.70498
  }
 }
 dps_results: {
  key: "TestBalance-Average-Default"
```
Comment (jimmyt857) [inline sim/druid/balance/TestBalance.results:902]: Looks like balance is losing 34 DPS in the average test. Is this expected?

File: sim/druid/talents.go [hunk 2]
```diff
@@ -314,14 +314,15 @@ func (druid *Druid) applyOmenOfClarity() {
 				spell.CostMultiplier += 1
 			}
 		},
-		OnCastComplete: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell) {
```
Comment (jimmyt857) [inline sim/druid/talents.go:317]: Instead of deleting this, I think we could have a local variable `var proccedAt time.Duration` which is set on proc, and then only remove the aura if `proccedAt != sim.CurrentTime`.

Then you'll be able to remove the hard-coded handling of clearcasting from all the spells.

File: sim/paladin/shield_of_righteousness.go [hunk 3]
```diff
@@ -9,6 +9,8 @@ import (
 
 func (paladin *Paladin) registerShieldOfRighteousnessSpell() {
 	baseCost := paladin.BaseMana * 0.06
+	has4pT8 := paladin.HasSetBonus(ItemSetAegisPlate, 4)
+	procAura := paladin.NewTemporaryStatsAura("Aegis", core.ActionID{SpellID: 64883}, stats.Stats{stats.BlockValue: 225}, time.Second*6)
```
Comment (jimmyt857) [inline sim/paladin/shield_of_righteousness.go:13]: Could we refactor this slightly like so?

```
var aegisPlateProcAura *core.Aura
if paladin.HasSetBonus(ItemSetAegisPlate, 4) {
  aegisPlateProcAura = paladin.NewTemporaryStatsAura("Aegis", core.ActionID{SpellID: 64883}, stats.Stats{stats.BlockValue: 225}, time.Second*6)
}

...
...

if aegisPlateProcAura != nil {
  aegisPlateProcAura.Activate(sim)
}
```

The advantage is that the aura won't be registered unless the set bonus is equipped, so in most cases we'll have 1 less aura attached to the unit.

File: sim/priest/items.go [hunk 4]
```diff
@@ -89,22 +89,22 @@ var ItemSetSanctificationRegalia = core.NewItemSet(core.ItemSet{
 		2: func(agent core.Agent) {
 		},
 		4: func(agent core.Agent) {
-			priest := agent.(PriestAgent).GetPriest()
```
Comment (jimmyt857) [inline sim/priest/items.go:92]: I think you can revert this change entirely. It seems likely that activating after the PWS cast is correct behavior.

File: sim/warlock/lifetap.go [hunk 5]
```diff
@@ -46,6 +47,14 @@ func (warlock *Warlock) registerLifeTapSpell() {
 			if hasGlyph {
 				warlock.GlyphOfLifeTapAura.Activate(sim)
 			}
+
+			if has4PT7 {
+				if warlock.SpiritsoftheDamnedAura.IsActive() {
```
Comment (jimmyt857) [inline sim/warlock/lifetap.go:52]: `aura.Activate()` already checks for `IsActive()` and does the refresh. You can simplify this code here, which will also cause the PPM metrics for the aura to be correct.

File: sim/warlock/items.go [hunk 6]
```diff
@@ -80,15 +80,15 @@ var ItemSetPlagueheartGarb = core.NewItemSet(core.ItemSet{
 				OnReset: func(aura *core.Aura, sim *core.Simulation) {
 					aura.Activate(sim)
 				},
-				OnCastComplete: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell) {
```
Comment (jimmyt857) [inline sim/warlock/items.go:83]: Seems like this change is a no-op? The new implementation in `life_tap.go` also applies the aura AFTER the calculation.

When reasonable, keeping the set bonus logic isolated to the set bonus function is preferred because its easier for the reader to understand.

File: sim/druid/balance/TestBalance.results (hunk 1 above)
Comment (Tharre) [reply sim/druid/balance/TestBalance.results:902]: If you look at the Remaining Differences commit, balance damage doesn't change at all there, so it should be fully explained by the caster trinket change.

File: sim/druid/talents.go (hunk 2 above)
Comment (Tharre) [reply sim/druid/talents.go:317]: I don't think that would work. The issue here why we have to change anything at all is that we have a CC proc -> we cast something for free -> OnSpellHitDealt is called and may proc another CC -> OnCastComplete is called and removes the CC proc. Weirdly even when moving the check into ApplyEffects the DPS is still different, I have no clue yet why that is.

Thinking about it, maybe we can check in OnCastComplete if the aura was applied at the exact same time as sim.CurrentTime and only remove it if they differ.

File: sim/paladin/shield_of_righteousness.go (hunk 3 above)
Comment (Tharre) [reply sim/paladin/shield_of_righteousness.go:13]: Yeah that makes more sense.

File: sim/priest/items.go (hunk 4 above)
Comment (Tharre) [reply sim/priest/items.go:92]: Agreed.

File: sim/warlock/lifetap.go (hunk 5 above)
Comment (Tharre) [reply sim/warlock/lifetap.go:52]: Oh, nice.

File: sim/warlock/items.go (hunk 6 above)
Comment (Tharre) [reply sim/warlock/items.go:83]: It is, yes. The only purpose of that commit was to prove that warlock DPS doesn't change from any other interaction.
